monitoring/ping.go

```diff
@@ -0,0 +1,189 @@
+type pingChecker struct {
+	serfRPCAddr string
+	serfRPCName string
+	rttStats    map[string]hdrhistogram.Histogram
+}
+
+// Name returns the checker name
+// Implements health.Checker
+func (c *pingChecker) Name() string {
+	return pingCheckerID
+}
+
+// Check verifies that all nodes' ping with Master Nodes is lower than the
+// desired threshold
+// Implements health.Checker
+func (c *pingChecker) Check(ctx context.Context, r health.Reporter) {
+	// FIXME: #1 RttThreshold will become configurable in future
+	// FIXME: #2 Send RttThreshold value to metrics
+
+	// fetch serf config and intantiate client
+	log.Debugf("Using Serf IP: %v", c.serfRPCAddr)
+	log.Debugf("Using Serf Name: %v", c.serfRPCName)
+	clientConfig := serf.Config{
+		Addr: c.serfRPCAddr,
+	}
+	client, err := serf.ClientFromConfig(&clientConfig)
+	if err != nil {
+		log.Errorf("error while connecting to Serf via IP %v. Error: %v",
+			c.serfRPCAddr, err.Error())
+		r.Add(&pb.Probe{
+			Checker: c.Name(),
+			Status:  pb.Probe_Failed,
+		})
+		return
+	}
+	defer client.Close()
+
+	// retrieve other nodes using Serf members
+	nodes, err := client.Members()
+	if err != nil {
+		log.Errorf("failed fetching Serf Members - %v", err.Error())
+		r.Add(&pb.Probe{
+			Checker: c.Name(),
+			Status:  pb.Probe_Failed,
+		})
+		return
+	}
+
+	// finding what is the current node
+	var self serf.Member
+	for _, node := range nodes {
+		// cut the port portion of the address after the ":" away
+		if node.Name == c.serfRPCName {
+			self = node
+		}
+	}
+	if self.Name == "" {
+		log.Errorf("error getting selfNode config: %s", c.serfRPCName)
+		r.Add(&pb.Probe{
+			Checker: c.Name(),
+			Status:  pb.Probe_Failed,
+		})
+		return
+	}
+
+	selfCoord, err := client.GetCoordinate(self.Name)
+	if err != nil || selfCoord == nil {
+		log.Errorf("error getting selfNode coordinates: %s -> %#v", self.Name, err)
+		r.Add(&pb.Probe{
+			Checker: c.Name(),
+			Status:  pb.Probe_Failed,
+		})
+		return
+	}
+	// ping each other node and fail in case the results are over a specified
+	// threshold
+	err = nil
+	for _, node := range nodes {
+		// skip pinging self
+		if node.Addr.String() == self.Addr.String() {
+			continue
+		}
+
+		coord2, err := client.GetCoordinate(node.Name)
+		if err != nil {
+			log.Errorf("error getting coordinates: %s -> %#v", node.Name, err)
+			r.Add(&pb.Probe{
+				Checker: c.Name(),
+				Status:  pb.Probe_Failed,
+			})
+			continue
+		}
+		if coord2 == nil {
+			err = trace.NotFound("could not find a coordinate for node %q", nodes[1])
+			log.Error(err)
+			r.Add(&pb.Probe{
+				Checker: c.Name(),
+				Status:  pb.Probe_Failed,
+			})
+			continue
+		}
+
+		// pingStats store ping statistics from 0 to 10000 ms (10 seconds)
+		// up to 3 digits precision
+		pingStats := hdrhistogram.New(0, 10000, 3)
+		for i := 0; i < slidingWindowSize; i++ {
+			rttNanoSec := selfCoord.DistanceTo(coord2).Nanoseconds()
+
+			if c.rttStats[node.Name] == nil { // FIXME: NOT WORKING
```

Comment:
You should initialize histograms when the pingChecker is created, then you won't need this.